EKF/gps_checks.cpp

```diff
@@ -0,0 +1,182 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file gps_checks.cpp
+ * Perform pre-flight and in-flight GPS quality checks
+ *
+ * @author Paul Riseborough <p_riseborough@live.com.au>
+ *
+ */
+
+#include "ekf.h"
+
+// GPS pre-flight check bit locations
+#define MASK_GPS_NSATS  (1<<0)
+#define MASK_GPS_GDOP   (1<<1)
+#define MASK_GPS_HACC   (1<<2)
+#define MASK_GPS_VACC   (1<<3)
+#define MASK_GPS_SACC   (1<<4)
+#define MASK_GPS_HDRIFT (1<<5)
+#define MASK_GPS_VDRIFT (1<<6)
+#define MASK_GPS_HSPD   (1<<7)
+#define MASK_GPS_VSPD   (1<<8)
+
+/*
+ * Return true if the GPS solution quality is adequate to set an origin for the EKF
+ * and start GPS aiding.
+ * All activated checks must pass for 10 seconds.
+ * Checks are activated using the EKF2_GPS_CHECKS bitmask parameter
+ * Checks are adjusted using the EKF2_REQ_* parameters
+*/
+bool EstimatorBase::gps_is_good(struct gps_message *gps)
+{
+    // Check the number of satellites
+    _gpsCheckFailStatus.flags.nsats = (gps->nsats < _params.requiredNsats);
+
+    // Check the geometric dilution of precision
+    _gpsCheckFailStatus.flags.gdop = (gps->gdop > _params.requiredGDoP);
+
+    // Check the reported horizontal position accuracy
+    _gpsCheckFailStatus.flags.hacc = (gps->eph > _params.requiredEph);
+
+    // Check the reported vertical position accuracy
+    _gpsCheckFailStatus.flags.vacc = (gps->epv > _params.requiredEpv);
+
+    // Check the reported speed accuracy
+    _gpsCheckFailStatus.flags.sacc = (gps->eph > _params.requiredEph);
+
+    // Calculate position movement since last measurement
+    float deltaPosN = 0.0f;
+    float deltaPosE = 0.0f;
+    double lat = gps->lat * 1.0e-7;
+    double lon = gps->lon * 1.0e-7;
+    if (_posRef.init_done) {
+        map_projection_project(&_posRef, lat, lon, &deltaPosN, &deltaPosE);
+    } else {
+        map_projection_init(&_posRef, lat, lon);
+        _gps_alt_ref = gps->alt * 1e-3f;
+    }
+
+    // Calculate time lapsesd since last update, limit to prevent numerical errors and calculate the lowpass filter coefficient
+    const float filtTimeConst = 10.0f;
+    float dt = fminf(fmaxf(float(_time_last_imu - _last_gps_origin_time_us)*1e-6f,0.001f),filtTimeConst);
+    float filterCoef = dt/filtTimeConst;
+
+    // Calculate the horizontal drift velocity components and limit to 10x the threshold
+    float velLimit = 10.0f * _params.requiredHdrift;
+    float velN = fminf(fmaxf(deltaPosN / dt, -velLimit), velLimit);
+    float velE = fminf(fmaxf(deltaPosE / dt, -velLimit), velLimit);
+
+    // Apply a low pass filter
+    _gpsDriftVelN = velN * filterCoef + _gpsDriftVelN * (1.0f - filterCoef);
+    _gpsDriftVelE = velE * filterCoef + _gpsDriftVelE * (1.0f - filterCoef);
+
+    // Calculate the horizontal drift speed and fail if too high
+    // This check can only be used if the vehicle is stationary during alignment
+    if(_vehicleArmed) {
```

Comment:
@priseborough Sorry to bother you about this but I think we should agree on code style. I tried to used google coding style until now
https://google.github.io/styleguide/cppguide.html#Variable_Names
Maybe we can discuss this sometime it's probably a good idea to do this at this stage.